File: 1370-count-number-of-nice-subarrays/count-number-of-nice-subarrays.java
import java.util.HashMap;

class Solution {
    public int numberOfSubarrays(int[] nums, int k) {
        int count = 0;
        int prefixSum = 0;
        HashMap<Integer, Integer> prefixCounts = new HashMap<>();
        prefixCounts.put(0, 1);

        for (int num : nums) {
            prefixSum += (num % 2 == 1) ? 1 : 0;
            if (prefixSum - k >= 0 && prefixCounts.containsKey(prefixSum - k)) {
                count += prefixCounts.get(prefixSum - k);
            }
            prefixCounts.put(prefixSum, prefixCounts.getOrDefault(prefixSum, 0) + 1);
        }

        return count;
    }
}

File: 1370-count-number-of-nice-subarrays/count-number-of-nice-subarrays.py
class Solution(object):
    def numberOfSubarrays(self, nums, k):
        from collections import defaultdict
    
        transformed = [1 if num % 2 == 1 else 0 for num in nums]
    
        count = 0
        prefix_sum = 0
        prefix_counts = defaultdict(int)
        prefix_counts[0] = 1
    
        for num in transformed:
            prefix_sum += num
            if prefix_sum - k in prefix_counts:
                count += prefix_counts[prefix_sum - k]
            prefix_counts[prefix_sum] += 1
    
        return count
